### PR TITLE
ITC-35 | Completed Prioritize Build Functionality

### PR DESCRIPTION
**Pull request by ShiviStar**
In this PR, Prioritize Build Functionality has been implemented.

### CODE REVIEW

File: app/components/repo-actions.js [hunk 1]
```diff
@@ -62,6 +63,7 @@ export default Component.extend({
   canCancel: and('userHasPullPermissionForRepo', 'item.canCancel'),
   canRestart: and('userHasPullPermissionForRepo', 'item.canRestart'),
   canDebug: and('userHasPushPermissionForRepo', 'item.canDebug'),
+  isQueued: reads('item.notStarted'),
```
Comment (artursmirnov) [inline app/components/repo-actions.js:66]: As far as I can see from the template code, there's no need to define an alias here, it's short enough to just use `item.notStarted` directly in template. What do you think?

File: app/templates/components/dialogs/prioritize-build-modal.hbs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    >
+      <SvgImage @name="icon-failed" @class="icon" />
+    </button>
+  </div>
+  <div class="repo-actions-modal__body">
+    <p>
+      Bypass your scheduled build by fast tracking this one.
+    </p>
+    <p class="label-align">
+      SELECT AN OPTION
```
Comment (artursmirnov) [inline app/templates/components/dialogs/prioritize-build-modal.hbs:25]: I'd like to suggest managing text casing via css `text-transform: uppercase` property, as soon as it's a styling concern.

File: app/templates/components/dialogs/prioritize-build-modal.hbs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        {{#let (eq this.selectionKey option.key) as |selected|}}
+          <button
+            class="repo-actions-item {{if selected 'repo-actions-item--selected'}}"
+            onclick={{action (mut this.selectionKey) option.key}}
+          >
+            <span class="repo-actions-item__radio"></span>
+            <span class="repo-actions-item__description">
+              {{option.description}}
+            </span>
+          </button>
```
Comment (artursmirnov) [inline app/templates/components/dialogs/prioritize-build-modal.hbs:38]: Would it be possible to re-use an existing https://github.com/travis-ci/travis-web/blob/master/app/components/forms/form-checkbox.js component here?

File: app/templates/components/dialogs/prioritize-build-modal.hbs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    </button>
+    {{#if (not this.isLoading)}}
+    <button
+      role="button"
+      class="button button--red"
+      {{on 'click' (fn @onClose)}}
+    >
+      Cancel
+    </button>
+    {{/if}}
```
Comment (artursmirnov) [inline app/templates/components/dialogs/prioritize-build-modal.hbs:64]: Indentation levels seems to be a bit messy here. Could you please indent the blocks so that nested ones were deeper and visualize hierarchy?

File: app/components/dialogs/prioritize-build-modal.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+import {
+  bindKeyboardShortcuts,
+  unbindKeyboardShortcuts
+} from 'ember-keyboard-shortcuts';
+
+export default Component.extend({
+  keyboardShortcuts: {
+    'esc': 'toggleConfirmationModal'
+  },
+  flashes: service(),
```
Comment (artursmirnov) [inline app/components/dialogs/prioritize-build-modal.js:16]: Please keep the spacing consistent across the entire file. Separate all blocks and properties by a single empty line, so that they were visually independent and easy to distinguish when scanning. The only case when empty line is not necessary is a group of semantically-similar single-line properties.

File: app/components/dialogs/prioritize-build-modal.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  }),
+
+  type: computed('job', 'build', function () {
+    let job = this.job;
+    if (job) {
+      return 'job';
+    } else {
+      return 'build';
+    }
+  }),
```
Comment (artursmirnov) [inline app/components/dialogs/prioritize-build-modal.js:66]: This could be just
```js
item: or('job', 'build'),
```
Isn't it?

File: app/components/dialogs/prioritize-build-modal.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+  initialKey: '',
+  initial: computed('initialKey', 'options.@each.key', function () {
+    return this.options.findBy('key', this.initialKey);
+  }),
+  initialIndex: computed('initial', 'options.[]', function () {
+    return this.options.indexOf(this.initial);
+  }),
```
Comment (artursmirnov) [inline app/components/dialogs/prioritize-build-modal.js:74]: I didn't find any use of these properties. Are they just a leftover?

File: app/templates/components/dialogs/prioritize-build-modal.hbs [hunk 8]
```diff
@@ -0,0 +1,66 @@
+<Modal
+  class="repo-actions-modal"
+  @isVisible={{@isOpen}}
```
Comment (artursmirnov) [inline app/templates/components/dialogs/prioritize-build-modal.hbs:3]: As long as there's a local `isOpen` property, defined on the component class, should it be used here instead of just reading attribute directly? E.g. `@isVisible={{this.isOpen}}`. Currently, it'll ignore the property on the component class.

File: app/components/dialogs/prioritize-build-modal.js [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  didInsertElement() {
+    this._super(...arguments);
+    bindKeyboardShortcuts(this);
+  },
+
+  willDestroyElement() {
+    this._super(...arguments);
+    unbindKeyboardShortcuts(this);
+  },
+  isOpen: false,
```
Comment (artursmirnov) [inline app/components/dialogs/prioritize-build-modal.js:27]: This property is not used because template uses component attribute directly by `@isOpen` - is it expected?

File: app/components/dialogs/prioritize-build-modal.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    return this.options.indexOf(this.initial);
+  }),
+
+  selectionKey: reads('initialKey'),
+  selection: computed('selectionKey', 'options.@each.key', function () {
+    return this.options.findBy('key', this.selectionKey);
+  }),
+  selectionIndex: computed('selection', 'options.[]', function () {
+    return this.options.indexOf(this.selection);
+  }),
```
Comment (artursmirnov) [inline app/components/dialogs/prioritize-build-modal.js:82]: This property is not used anywhere. Is it a leftover?

File: app/components/dialogs/prioritize-build-modal.js [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+      {
+        key: 'private2',
+        displayValue: 'item2',
+        cancelRunningJobsVal: true,
+        description: 'Place build at the top of the queue and cancel all running jobs',
+        modalText: 'Place build at the top of the queue and cancel all running jobs',
+      }
+    ];
+    return arr;
+  }),
```
Comment (artursmirnov) [inline app/components/dialogs/prioritize-build-modal.js:46]: All this data is mainly presentation. I'd like to suggest moving it to template, as presentation is the template's concern. Also, as long as there are just two options currently, there's no need for iteration, just put them both one by one, set the `cancelRunningJobsVal` on change and calculate selection state by equality to this property, e.g. `selected={{eq cancelRunningJobsVal true}}` and `selected={{eq cancelRunningJobsVal false}}` accordingly. Does it make sense to you?

File: app/components/dialogs/prioritize-build-modal.js [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+      }, () => {
+        this.flashes.error('An error occurred. The build could not be prioritized.');
+      });
+    });
+  }).drop(),
+  actions: {
+    toggleConfirmationModal() {
+      if (!this.isLoading) {
+        this.set('isOpen', false);
+        this.set('doAutofocus', true);
```
Comment (artursmirnov) [inline app/components/dialogs/prioritize-build-modal.js:100]: I didn't find any use of `doAutofocus`. What that is?

File: app/components/dialogs/prioritize-build-modal.js [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        this.flashes.success('The build was successfully prioritized.');
+        this.set('isLoading', false);
+        this.set('isOpen', false);
+      }, () => {
+        this.flashes.error('An error occurred. The build could not be prioritized.');
+      });
+    });
+  }).drop(),
+  actions: {
+    toggleConfirmationModal() {
```
Comment (artursmirnov) [inline app/components/dialogs/prioritize-build-modal.js:97]: Since this action seems to only close the modal, would it make sense to rename it accordingly?

File: app/components/dialogs/prioritize-build-modal.js [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+  selectionKey: reads('initialKey'),
+  selection: computed('selectionKey', 'options.@each.key', function () {
+    return this.options.findBy('key', this.selectionKey);
+  }),
+  selectionIndex: computed('selection', 'options.[]', function () {
+    return this.options.indexOf(this.selection);
+  }),
+
+  increasePriorityTask: task(function* (value) {
+    this.set('isLoading', true);
```
Comment (artursmirnov) [inline app/components/dialogs/prioritize-build-modal.js:85]: Please use `task`'s [derived state](http://ember-concurrency.com/docs/derived-state) to get loading state consistently.

File: app/components/dialogs/prioritize-build-modal.js [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    yield eventually(this.item, (record) => {
```
Comment (artursmirnov) [inline app/components/dialogs/prioritize-build-modal.js:86]: I don't think there's a need to use `eventually` here, because at the moment when this modal is shown, the build/job should already be resolved. What do you think?

File: app/components/dialogs/prioritize-build-modal.js [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+      record.increasePriority(this.selection.cancelRunningJobsVal).then((response) => {
+        this.flashes.success('The build was successfully prioritized.');
+        this.set('isLoading', false);
+        this.set('isOpen', false);
+      }, () => {
+        this.flashes.error('An error occurred. The build could not be prioritized.');
+      });
```
Comment (artursmirnov) [inline app/components/dialogs/prioritize-build-modal.js:93]: Please avoid using `then`/`catch` (especially with callbacks signature) in tasks, as they are generator functions, and using promises inside of them just gets rid of all advantages they provide:
```js
try {
  yield record.increasePriority(this.selection.cancelRunningJobsVal);
  this.flashes.success('The build was successfully prioritized.');
  this.set('isOpen', false);
} catch (error) {
  this.flashes.error('An error occurred. The build could not be prioritized.');
}
```

File: app/models/job.js [hunk 17]
```diff
@@ -243,5 +243,9 @@ export default Model.extend(DurationCalculations, DurationAttributes, {
   didLoad() {
     if (this.number)
       this.set('jobIdNumber', this.number);
-  }
+  },
+  increasePriority(cancellAllRunningJobs) {
+    const url = `/build/${this.id}/priority?${cancellAllRunningJobs}`;
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/models/job.js:248]: Done. Removed function from job.js and used build function.

File: app/styles/app/modules/buttons.scss [hunk 18]
```diff
@@ -233,7 +233,8 @@
 
 .action-button--restart,
 .action-button--debug,
-.action-button--cancel {
+.action-button--cancel,
+.action-button--Priortize {
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/styles/app/modules/buttons.scss:237]: Done.

File: app/styles/app/modules/buttons.scss [hunk 19]
```diff
@@ -257,6 +258,10 @@
   }
 }
 
+.action-button--Priortize{
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/styles/app/modules/buttons.scss:261]: Done.

File: app/styles/app/modules/prioritize-build.scss [hunk 20]
```diff
@@ -0,0 +1,97 @@
+.repo-actions {
+    list-style: none;
+    padding: .5rem 1rem;
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/styles/app/modules/prioritize-build.scss:3]: Done.

File: app/styles/app/modules/prioritize-build.scss [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+    font-style: italic;
+    color: $cement-grey;
+    background-color: #f5fbfb;
+    border-radius: 2px;
+  
+    &-item {
+      justify-content: flex-start;
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/styles/app/modules/prioritize-build.scss:10]: Removed this property.

File: app/styles/app/modules/prioritize-build.scss [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+    }
+  
+    & &__close-button {
+      margin: 0;
+      padding: 0;
+      outline: none;
+    }
+  
+    & &__header {
+      font-size: 22px;
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/styles/app/modules/prioritize-build.scss:71]: Done.

File: app/styles/app/modules/prioritize-build.scss [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+  
+    & &__body {
+      color: $asphalt-grey;
+    }
+  
+    & &__controls {
+      .loading-indicator {
+        float: left;
+      }
+      .button {
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/styles/app/modules/prioritize-build.scss:85]: Done.

File: app/styles/app/modules/prioritize-build.scss [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+        margin: 0 1rem 0 0;
+        display: inline-block;
+        width: auto;
+        height: auto;
+        padding: 9px;
+        margin-right: 5px;
+        border-radius: 2px;
+        border-color: #9d9d9d;
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/styles/app/modules/prioritize-build.scss:93]: Done.

File: app/templates/components/build-header.hbs [hunk 25]
```diff
@@ -183,6 +183,10 @@
           <SvgImage @name="icon-shared" @alt="Shared repository" @class="icon shared" /> Shared repository
         {{/if}}
       {{/if}}
+      {{#if this.build.priority}}
+        <SvgImage @name="icon-priortize" @class="icon" />
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/templates/components/build-header.hbs:187]: Done.

File: app/templates/components/build-header.hbs [hunk 26]
```diff
@@ -183,6 +183,10 @@
           <SvgImage @name="icon-shared" @alt="Shared repository" @class="icon shared" /> Shared repository
         {{/if}}
       {{/if}}
+      {{#if this.build.priority}}
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/templates/components/build-header.hbs:186]: @artursmirnov: 'priority' field is added by the backend team in the build api to calculate the highest priority build. Its value will be in boolean format.

File: app/templates/components/repo-actions.hbs [hunk 27]
```diff
@@ -1,3 +1,10 @@
+<Dialogs::PrioritizeBuildModal 
+  @isOpen={{this.showCancelModal}}
+  @onClose={{fn (mut this.showCancelModal) false}}
+  @increasePriority={{this.increasePriority}}
+  @build={{this.build}}
+  @job={{this.job}}
+/>
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/templates/components/repo-actions.hbs:7]: Done.

File: app/templates/components/repo-actions.hbs [hunk 28]
```diff
@@ -21,6 +28,31 @@
         </span>
       </button>
     </div>
+{{/if}}
+  {{#if (and (not this.build.priority) this.build.permissions.prioritize this.isQueued)}}
+    <div class="action-button-container">
+        {{#if (not this.userHasPushPermissionForRepo)}}
+          <EmberTooltip
+          class={{if this.mobilelabels "hidden-on-mobile"}} 
+          @text="This action can be triggered only by users with push access to the repository" />
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/templates/components/repo-actions.hbs:37]: Done.

File: app/templates/components/repo-actions.hbs [hunk 29]
```diff
@@ -21,6 +28,31 @@
...
+        {{/if}}
+      <button 
+        type="button"
+        class="action-button--Priortize"
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/templates/components/repo-actions.hbs:41]: Done.

File: app/templates/components/repo-actions.hbs [hunk 30]
```diff
@@ -1,3 +1,10 @@
+<Dialogs::PrioritizeBuildModal 
+  @isOpen={{this.showCancelModal}}
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/templates/components/repo-actions.hbs:2]: Done.

File: app/templates/components/repo-actions.hbs [hunk 31]
```diff
@@ -21,6 +28,31 @@
         </span>
       </button>
     </div>
+{{/if}}
+  {{#if (and (not this.build.priority) this.build.permissions.prioritize this.isQueued)}}
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/templates/components/repo-actions.hbs:32]: Already explained in above comment.

File: app/templates/components/repo-actions.hbs [hunk 32]
```diff
@@ -21,6 +28,31 @@
...
+        class="action-button--Priortize"
+        {{on 'click' (fn (mut this.showCancelModal) true)}}
+        disabled={{not this.userHasPushPermissionForRepo}}
+      >
+        {{#if this.tooltips}}
+          <EmberTooltip class={{if this.mobilelabels "hidden-on-mobile"}}>
+            Prioritize build
+          </EmberTooltip>
+        {{/if}}
+        <SvgImage @name="icon-priortize" @class="icon" />
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/templates/components/repo-actions.hbs:50]: Done.

File: app/components/repo-actions.js [hunk 33]
```diff
@@ -1,14 +1,15 @@
...
-import { alias, and, or } from '@ember/object/computed';
+import { alias, and, or, reads } from '@ember/object/computed';
 import eventually from 'travis/utils/eventually';
 import { task, taskGroup } from 'ember-concurrency';
 
 export default Component.extend({
   flashes: service(),
   features: service(),
   auth: service(),
+  api: service(),
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/components/repo-actions.js:12]: Removed it.

File: app/components/repo-actions.js (hunk 1 above)
Comment (ShiviStar) [reply app/components/repo-actions.js:66]: Done.

File: app/templates/components/dialogs/prioritize-build-modal.hbs (hunk 2 above)
Comment (ShiviStar) [reply app/templates/components/dialogs/prioritize-build-modal.hbs:25]: Done.

File: app/templates/components/dialogs/prioritize-build-modal.hbs (hunk 4 above)
Comment (ShiviStar) [reply app/templates/components/dialogs/prioritize-build-modal.hbs:64]: Done.

File: app/components/dialogs/prioritize-build-modal.js (hunk 5 above)
Comment (ShiviStar) [reply app/components/dialogs/prioritize-build-modal.js:16]: Done.

File: app/components/dialogs/prioritize-build-modal.js (hunk 6 above)
Comment (ShiviStar) [reply app/components/dialogs/prioritize-build-modal.js:66]: Done.

File: app/components/dialogs/prioritize-build-modal.js (hunk 7 above)
Comment (ShiviStar) [reply app/components/dialogs/prioritize-build-modal.js:74]: Removed.

File: app/templates/components/dialogs/prioritize-build-modal.hbs (hunk 8 above)
Comment (ShiviStar) [reply app/templates/components/dialogs/prioritize-build-modal.hbs:3]: Used {{this.isOpen}} instead of {{@isOpen}} as suggested.

File: app/components/dialogs/prioritize-build-modal.js (hunk 9 above)
Comment (ShiviStar) [reply app/components/dialogs/prioritize-build-modal.js:27]: Removed now.

File: app/components/dialogs/prioritize-build-modal.js (hunk 10 above)
Comment (ShiviStar) [reply app/components/dialogs/prioritize-build-modal.js:82]: Removed.

File: app/components/dialogs/prioritize-build-modal.js (hunk 12 above)
Comment (ShiviStar) [reply app/components/dialogs/prioritize-build-modal.js:100]: No, it is of no use. Removed it.

File: app/components/dialogs/prioritize-build-modal.js (hunk 13 above)
Comment (ShiviStar) [reply app/components/dialogs/prioritize-build-modal.js:97]: Renamed it.

File: app/templates/components/dialogs/prioritize-build-modal.hbs (hunk 3 above)
Comment (ShiviStar) [reply app/templates/components/dialogs/prioritize-build-modal.hbs:38]: @artursmirnov: The target here is to implement radio buttons but the link which you have provided is for checkbox.

Comment (artursmirnov) [reply app/templates/components/dialogs/prioritize-build-modal.hbs:38]: @ShiviStar, you're right, it's checkbox, but it looks and works very much in the same way:

2020-06-16 at 16 23

And it's easy to achieve this behavior:

```js
<form.field
  @disableFrame={{true}}
  @value={{eq this.shouldCancel true}}
  @onChange={{action (mut this.shouldCancel) true}}
  as |field|
>
  <field.checkbox>
    Cancel builds
  </field.checkbox>
</form.field>

<form.field
  @disableFrame={{true}}
  @value={{eq this.shouldCancel false}}
  @onChange={{action (mut this.shouldCancel) false}}
  as |field|
>
  <field.checkbox>
    Don't cancel builds
  </field.checkbox>
</form.field>
```
Does it make sense?

Please note that the code above is just an example with fake properties and texts

File: app/components/dialogs/prioritize-build-modal.js (hunk 15 above)
Comment (ShiviStar) [reply app/components/dialogs/prioritize-build-modal.js:86]: @artursmirnov : May you please let me know how I can interact with 'build' model in this case as in current scenario I am passing this.item in eventually function which is helping me to interact with model methods.

Comment (artursmirnov) [reply app/components/dialogs/prioritize-build-modal.js:86]: @ShiviStar, sure! The only thing `eventually` does is it resolves a record if it has not been yet resolved. But _theoretically_ at the moment when this task is getting executed, the model should already be resolved, because it is rendered on current view. Therefore, `this.item` should contain already resolved record for either `build` or `job`, depending on current context. Although, you have the `increasePriority` method only for `build`, so you should treat them differently here: either `this.item.increasePriority()` for build or `this.item.build. increasePriority()` for job. 

Does it make sense?

File: app/templates/components/build-header.hbs (hunk 26 above)
[the remark this answers is not in the document]
Comment (artursmirnov) [reply app/templates/components/build-header.hbs:186]: I thought it would be the case, but in order for this to work, this field should also be added to the `build` model as attribute. Otherwise it will never appear on records and will not be available here.

File: app/styles/app/modules/prioritize-build.scss [hunk 34]
```diff
@@ -0,0 +1,97 @@
...
+      justify-content: flex-start;
+      align-items: baseline;
+      margin: .5rem 0;
+      padding: 0;
+      background-color: transparent;
+      cursor: pointer;
+      outline: none;
+      display: block !important;
+      transform: none !important;
+      width: auto !important;
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/styles/app/modules/prioritize-build.scss:18]: Removed.

File: app/templates/components/build-header.hbs (hunk 26 above)
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/templates/components/build-header.hbs:186]: Added the key in build model as well.

File: app/models/build.js [hunk 35]
```diff
@@ -159,4 +161,9 @@ export default Model.extend(DurationCalculations, {
       return m.isValid() ? m.format('lll') : 'not finished yet';
     }
   }),
+
+  increasePriority(cancellAllRunningJobs) {
+    const url = `/build/${this.id}/priority?${cancellAllRunningJobs}`;
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/models/build.js]: Handled the validation. Now the variable always contains a value.

File: app/styles/app/modules/buttons.scss [hunk 36]
```diff
@@ -325,13 +330,13 @@
...
+  //   transform: translate(-5px, -2px);
 
-    button {
-      background: transparent;
-    }
-  }
+  //   button {
+  //     background: transparent;
+  //   }
+  // }
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/styles/app/modules/buttons.scss]: I have checked and this is not breaking any existing functionality.

File: app/templates/components/dialogs/prioritize-build-modal.hbs (hunk 3 above)
Comment (RamMount) [reply app/templates/components/dialogs/prioritize-build-modal.hbs]: Implemented as suggested.

Comment (ShiviStar) [reply app/templates/components/dialogs/prioritize-build-modal.hbs]: Implemented as suggested.

File: app/components/dialogs/prioritize-build-modal.js (hunk 11 above)
Comment (ShiviStar) [reply app/components/dialogs/prioritize-build-modal.js]: Implemented as suggested.

File: app/components/dialogs/prioritize-build-modal.js (hunk 5 above)
Comment (artursmirnov) [reply app/components/dialogs/prioritize-build-modal.js:17]: Perfect, thank you!

File: app/components/repo-actions.js [hunk 37]
```diff
@@ -1,7 +1,7 @@
 import Component from '@ember/component';
 import { inject as service } from '@ember/service';
 import { computed } from '@ember/object';
-import { alias, and, or } from '@ember/object/computed';
+import { alias, and, or, reads } from '@ember/object/computed';
```
Comment (artursmirnov) [inline app/components/repo-actions.js:4]: This change seems to be unnecessary
```suggestion
import { alias, and, or} from '@ember/object/computed';
```

File: app/models/build.js [hunk 38]
```diff
@@ -159,4 +163,9 @@ export default Model.extend(DurationCalculations, {
       return m.isValid() ? m.format('lll') : 'not finished yet';
     }
   }),
+
+  increasePriority(shouldCancelRunningJobs) {
+    const url = `/build/${this.id}/priority?cancel_all=${shouldCancelRunningJobs}`;
```
Comment (artursmirnov) [inline app/models/build.js:168]: I would like to suggest making this url a little bit more predictable:
```suggestion
    const url = `/build/${this.id}/priority?cancel_all=${shouldCancelRunningJobs ? true : false}`;
```
This would prevent all the `null`, `undefined`, etc. to be assigned to the URL.

File: app/styles/app/modules/buttons.scss [hunk 39]
```diff
@@ -233,7 +233,8 @@
 
 .action-button--restart,
 .action-button--debug,
-.action-button--cancel {
+.action-button--cancel,
+.action-button--priortize {
```
Comment (artursmirnov) [inline app/styles/app/modules/buttons.scss:237]: There's a typo in the class name:
```suggestion
.action-button--prioritize {
```

File: app/styles/app/modules/buttons.scss [hunk 40]
```diff
@@ -257,6 +258,10 @@
   }
 }
 
+.action-button--priortize{
```
Comment (artursmirnov) [inline app/styles/app/modules/buttons.scss:261]: There's a typo in the class name:
```suggestion
.action-button--prioritize {
```

File: app/styles/app/modules/buttons.scss [hunk 41]
```diff
@@ -326,11 +331,11 @@
...
   .repo-main-tools {
-    transform: translate(-5px, -2px);
+    // transform: translate(-5px, -2px);
 
-    button {
-      background: transparent;
-    }
+    // button {
+    //   background: transparent;
+    // }
```
Comment (artursmirnov) [inline app/styles/app/modules/buttons.scss:338]: Please remove the commented out code in case if it's not needed anymore

File: app/templates/components/build-header.hbs (hunk 26 above)
[the remark this answers is not in the document]
Comment (artursmirnov) [reply app/templates/components/build-header.hbs:186]: Thank you 👍

File: app/components/dialogs/prioritize-build-modal.js [hunk 42]
```diff
@@ -0,0 +1,51 @@
...
+        this.flashes.success('The build was successfully prioritized.');
+        this.set('isOpen', false);
+      }, () => {
+        this.flashes.error('An error occurred. The build could not be prioritized.');
+      });
+    });
+  }).drop(),
+
+  actions: {
+    toggleCloseModal() {
```
Comment (artursmirnov) [inline app/components/dialogs/prioritize-build-modal.js:45]: How does `closeModal` sound to you?
```suggestion
    closeModal() {
```

File: app/components/dialogs/prioritize-build-modal.js [hunk 43]
```diff
@@ -0,0 +1,51 @@
...
+
+  increasePriorityTask: task(function* () {
+    yield eventually(this.item, (record) => {
+      record.increasePriority(this.shouldCancelRunningJobs).then(() => {
+        this.flashes.success('The build was successfully prioritized.');
+        this.set('isOpen', false);
+      }, () => {
+        this.flashes.error('An error occurred. The build could not be prioritized.');
+      });
+    });
```
Comment (artursmirnov) [inline app/components/dialogs/prioritize-build-modal.js:41]: @ShiviStar the `task`'s `isRunning` property doesn't reflect the expected value because `eventually` does not return a promise, therefore, the `task` is not even involved in the asynchrony, it just finishes synchronously and doesn't know when the request is done.

Could you please try this code:
```js
    try {
      const record = yield this.item; // this will do the same as `eventually` - resolve the record
      yield record.increasePriority(this.shouldCancelRunningJobs); // this should work for `build`, but won't work for `job` because `job` doesn't have `increasePriority` method
      this.flashes.success('The build was successfully prioritized.');
      this.set('isOpen', false);
    } catch (error) {
      this.flashes.error('An error occurred. The build could not be prioritized.');
    }
```
With this code it should work properly. Please let me know if it doesn't.

**Comment by artursmirnov** [comment]
@ShiviStar I have tested the branch visually and it looks very good! Great job, thank you! The only thing that caught my eye is the button alignment:

image

Could you please align it by the right side and also increase the right padding for the button so that the end of the title aligns with the "Cancel build" button?

File: app/components/dialogs/prioritize-build-modal.js (hunk 43 above)
Comment (Rahulgarg30591) [reply app/components/dialogs/prioritize-build-modal.js]: Hello, @artursmirnov thank you for assisting in this, and yes it is working and resolves both our open issues of removal of eventually and consuming isRunning state. Thank you again.

File: app/components/dialogs/prioritize-build-modal.js (hunk 15 above)
Comment (Rahulgarg30591) [reply app/components/dialogs/prioritize-build-modal.js]: Done.

File: app/components/dialogs/prioritize-build-modal.js (hunk 16 above)
Comment (Rahulgarg30591) [reply app/components/dialogs/prioritize-build-modal.js]: Implemented as suggested. Now done.

File: app/models/build.js (hunk 38 above)
Comment (Rahulgarg30591) [reply app/models/build.js:168]: Hello, @artursmirnov it is giving _no-unneeded-ternary_ lint issue.

File: app/components/repo-actions.js (hunk 37 above)
Comment (Rahulgarg30591) [reply app/components/repo-actions.js:4]: Removed.

File: app/styles/app/modules/buttons.scss (hunk 39 above)
Comment (Rahulgarg30591) [reply app/styles/app/modules/buttons.scss:237]: Done.

File: app/styles/app/modules/buttons.scss (hunk 40 above)
Comment (Rahulgarg30591) [reply app/styles/app/modules/buttons.scss:261]: Done.

File: app/models/build.js (hunk 38 above)
Comment (Rahulgarg30591) [reply app/models/build.js]: Hence now implementing it with explicit if conditions.

File: app/components/dialogs/prioritize-build-modal.js (hunk 14 above)
Comment (Rahulgarg30591) [reply app/components/dialogs/prioritize-build-modal.js]: Done.

File: app/styles/app/modules/buttons.scss (hunk 41 above)
Comment (Rahulgarg30591) [reply app/styles/app/modules/buttons.scss]: Done.

File: app/components/dialogs/prioritize-build-modal.js (hunk 42 above)
Comment (Rahulgarg30591) [reply app/components/dialogs/prioritize-build-modal.js]: Done.

File: app/components/dialogs/prioritize-build-modal.js (hunk 43 above)
Comment (artursmirnov) [reply app/components/dialogs/prioritize-build-modal.js]: Happy to help! :)

**Comment by ShiviStar** [comment]
@artursmirnov The Build buttons alignment are fixed now. Kindly review it.

File: app/models/build.js [hunk 44]
```diff
@@ -13,6 +13,8 @@ export default Model.extend(DurationCalculations, {
 
   branchName: alias('branch.name'),
 
+  permissions: attr(),
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/models/build.js]: @artursmirnov I have removed this change from my branch in the latest commit.

File: app/serializers/build_v2_fallback.js [hunk 45]
```diff
@@ -37,7 +37,11 @@ export default V2FallbackSerializer.extend({
   normalize: function (modelClass, resourceHash) {
     // TODO: remove this after switching to V3 entirely
     let type = resourceHash['@type'];
+    let permissionsHash = resourceHash['@permissions'];
     let commit = resourceHash.commit;
+    if (permissionsHash) {
+      resourceHash.permissions = permissionsHash;
+    }
```
[the remark this answers is not in the document]
Comment (ShiviStar) [reply app/serializers/build_v2_fallback.js]: @artursmirnov I have removed this change from my branch in the latest commit.

File: app/serializers/build_v2_fallback.js [hunk 46]
```diff
@@ -37,11 +37,7 @@ export default V2FallbackSerializer.extend({
   normalize: function (modelClass, resourceHash) {
     // TODO: remove this after switching to V3 entirely
     let type = resourceHash['@type'];
-    let permissionsHash = resourceHash['@permissions'];
     let commit = resourceHash.commit;
-    if (permissionsHash) {
-      resourceHash.permissions = permissionsHash;
-    }
```
Comment (artursmirnov) [inline app/serializers/build_v2_fallback.js:44]: Could you please revert this change? My previous comments for this change was to pull it from the upstream (which you've already done) instead of re-introducing it in the branch. With these lines removed the `permissions` property will never work on `build` model.

Comment (ShiviStar) [reply app/serializers/build_v2_fallback.js]: Done.

Comment (artursmirnov) [reply app/serializers/build_v2_fallback.js]: Thank you! 👍

**Comment by ShiviStar** [comment]
@artursmirnov @shairyar Thank you for reviewing and assisting us.
